File: agent-skill/stackprism-site-experience/scripts/bridge/security.mjs
import { fail, isValidId } from './protocol.mjs'

const duplicateSensitiveHeaders = new Set(['host', 'authorization', 'content-type', 'content-length'])

const hasDuplicateSensitiveHeaders = rawHeaders => {
  const seen = new Set()
  for (let index = 0; index < rawHeaders.length; index += 2) {
    const name = String(rawHeaders[index] || '').toLowerCase()
    if (!duplicateSensitiveHeaders.has(name)) continue
    if (seen.has(name)) return true
    seen.add(name)
  }
  return false
}

export const rejectBadRequestShell = (req, res, baseUrl) => {
  if (hasDuplicateSensitiveHeaders(req.rawHeaders || [])) {
    fail(res, 400, 'INVALID_REQUEST', 'Ambiguous request headers are not allowed.')
    return true
  }
  if (req.headers.host !== new URL(baseUrl).host) {
    fail(res, 400, 'INVALID_REQUEST', 'Host is not allowed.')
    return true
  }
  const target = String(req.url || '')
  if (!target.startsWith('/') || target.startsWith('//')) {
    fail(res, 400, 'INVALID_REQUEST', 'Only origin-form request targets are allowed.')
    return true
  }
  const [rawPath, rawQuery = ''] = target.split('?', 2)
  if (/%2e|%2f|%5c|\\/i.test(target)) {
    fail(res, 400, 'INVALID_REQUEST', 'Encoded path separators or dot segments are not allowed.')
    return true
  }
  if (
    rawPath !== '/' &&
    rawPath
      .split('/')
      .slice(1)
      .some(segment => segment === '' || segment === '.' || segment === '..')
  ) {
    fail(res, 400, 'INVALID_REQUEST', 'Ambiguous path segments are not allowed.')
    return true
  }
  if (rawQuery && rawPath !== '/bridge') {
    fail(res, 400, 'INVALID_REQUEST', 'Query string is not allowed for this endpoint.')
    return true
  }
  const contentLength = req.headers['content-length']
  if (contentLength && !/^\d+$/.test(String(contentLength))) {
    fail(res, 400, 'INVALID_REQUEST', 'Content-Length is invalid.')
    return true
  }
  const contentEncoding = req.headers['content-encoding']
  if (contentEncoding && String(contentEncoding).toLowerCase() !== 'identity') {
    fail(res, 415, 'UNSUPPORTED_MEDIA_TYPE', 'Content-Encoding is not supported.')
    return true
  }
  if (req.headers['transfer-encoding'] && req.headers['content-length']) {
    fail(res, 400, 'INVALID_REQUEST', 'Content-Length and Transfer-Encoding cannot be combined.')
    return true
  }
  if (req.headers['transfer-encoding']) {
    fail(res, 400, 'UNSUPPORTED_TRANSFER_ENCODING', 'Transfer-Encoding is not supported.')
    return true
  }
  return false
}

export const rejectCrossOriginSensitiveRequest = (req, res, baseUrl) => {
  let allowedOrigin = baseUrl
  try {
    allowedOrigin = new URL(baseUrl).origin
  } catch {}

  const { origin } = req.headers
  if (origin && origin !== allowedOrigin) {
    fail(res, 403, 'ORIGIN_NOT_ALLOWED', 'Origin is not allowed.')
    return true
  }

  const { referer } = req.headers
  if (referer) {
    try {
      if (new URL(referer).origin !== allowedOrigin) {
        fail(res, 403, 'ORIGIN_NOT_ALLOWED', 'Referer is not allowed.')
        return true
      }
    } catch {
      fail(res, 403, 'ORIGIN_NOT_ALLOWED', 'Referer is not allowed.')
      return true
    }
  }

  const fetchSite = req.headers['sec-fetch-site']
  if (fetchSite && !['same-origin', 'none'].includes(fetchSite)) {
    fail(res, 403, 'ORIGIN_NOT_ALLOWED', 'Sec-Fetch-Site is not allowed.')
    return true
  }
  return false
}

const bodyTimeout = Symbol('bodyTimeout')

export const readJson = async (req, limit = 5 * 1024 * 1024, timeoutMs = 0) => {
  if (!/^application\/json(?:;\s*charset=utf-8)?$/i.test(req.headers['content-type'] || '')) {
    return { ok: false, status: 415, code: 'UNSUPPORTED_MEDIA_TYPE', message: 'Expected application/json.' }
  }
  const chunks = []
  let size = 0
  const readBody = (async () => {
    for await (const chunk of req) {
      size += chunk.byteLength
      if (size > limit) return { ok: false, status: 413, code: 'REQUEST_TOO_LARGE', message: 'Request body is too large.', close: true }
      chunks.push(chunk)
    }
    return { ok: true }
  })()
  readBody.catch(() => {})
  let bodyResult
  let timeoutId
  try {
    const timeout =
      timeoutMs > 0
        ? new Promise(resolve => {
            timeoutId = setTimeout(() => resolve(bodyTimeout), timeoutMs)
          })
        : null
    bodyResult = timeout ? await Promise.race([readBody, timeout]) : await readBody
  } catch {
    return { ok: false, status: 400, code: 'INVALID_JSON', message: 'Request body is not valid JSON.' }
  } finally {
    if (timeoutId) clearTimeout(timeoutId)
  }
  if (bodyResult === bodyTimeout) {
    return { ok: false, status: 408, code: 'REQUEST_TIMEOUT', message: 'Request body timed out.', close: true }
  }
  if (!bodyResult.ok) return bodyResult
  try {
    const text = new TextDecoder('utf-8', { fatal: true }).decode(Buffer.concat(chunks))
    return { ok: true, body: JSON.parse(text) }
  } catch {
    return { ok: false, status: 400, code: 'INVALID_JSON', message: 'Request body is not valid JSON.' }
  }
}

const bridgeQueryKinds = {
  session: 'sessionId',
  capture: 'captureId',
  nonce: 'nonce'
}

export const parseBridgeQuery = rawSearch => {
  const raw = String(rawSearch || '').replace(/^\?/, '')
  const parts = raw ? raw.split('&') : []
  if (parts.length !== 3) return null
  const values = {}
  for (const part of parts) {
    const separatorIndex = part.indexOf('=')
    if (!part || separatorIndex <= 0 || part.indexOf('=', separatorIndex + 1) !== -1) return null
    const name = part.slice(0, separatorIndex)
    const value = part.slice(separatorIndex + 1)
    const kind = bridgeQueryKinds[name]
    if (!kind || Object.prototype.hasOwnProperty.call(values, name) || !isValidId(kind, value)) return null
    values[name] = value
  }
  return Object.keys(values).length === 3 ? values : null
}

export const validateBridgeQuery = url => Boolean(parseBridgeQuery(url.search))

File: agent-skill/stackprism-site-experience/scripts/stackprism_bridge_lib/security.py
from urllib.parse import urlparse

from .protocol import safe_equal, valid_id


SENSITIVE_DUPLICATE_HEADERS = {"Host", "Authorization", "Content-Type", "Content-Length"}
BRIDGE_QUERY_KINDS = {
    "session": "sessionId",
    "capture": "captureId",
    "nonce": "nonce",
}


def parse_bridge_query(parsed_query):
    parts = parsed_query.split("&") if parsed_query else []
    if len(parts) != 3:
        return None
    values = {}
    for part in parts:
        if not part or part.count("=") != 1:
            return None
        name, value = part.split("=", 1)
        kind = BRIDGE_QUERY_KINDS.get(name)
        if not kind or name in values or not valid_id(kind, value):
            return None
        values[name] = value
    return values if set(values.keys()) == set(BRIDGE_QUERY_KINDS.keys()) else None


def valid_bridge_query(parsed_query):
    return parse_bridge_query(parsed_query) is not None


def bridge_query_value(parsed_query, name):
    return (parse_bridge_query(parsed_query) or {}).get(name, "")


def bad_shell_error(handler):
    for name in SENSITIVE_DUPLICATE_HEADERS:
        if len(handler.headers.get_all(name, [])) > 1:
            return 400, "INVALID_REQUEST", "Ambiguous request headers are not allowed."
    content_length = handler.headers.get("Content-Length")
    if content_length is not None and (not content_length.isdecimal()):
        return 400, "INVALID_REQUEST", "Content-Length is invalid."
    if handler.headers.get("Host") != urlparse(handler.server.store.base_url).netloc:
        return 400, "INVALID_REQUEST", "Host is not allowed."
    path = handler.path
    if not path.startswith("/") or path.startswith("//"):
        return 400, "INVALID_REQUEST", "Only origin-form request targets are allowed."
    raw_path, _, raw_query = path.partition("?")
    if any(value in path.lower() for value in ("%2e", "%2f", "%5c")) or "\\" in path:
        return 400, "INVALID_REQUEST", "Encoded path separators or dot segments are not allowed."
    if raw_path != "/" and any(segment in {"", ".", ".."} for segment in raw_path.split("/")[1:]):
        return 400, "INVALID_REQUEST", "Ambiguous path segments are not allowed."
    if raw_query and raw_path != "/bridge":
        return 400, "INVALID_REQUEST", "Query string is not allowed for this endpoint."
    content_encoding = handler.headers.get("Content-Encoding")
    if content_encoding and content_encoding.lower() != "identity":
        return 415, "UNSUPPORTED_MEDIA_TYPE", "Content-Encoding is not supported."
    transfer_encoding = handler.headers.get("Transfer-Encoding")
    if transfer_encoding and handler.headers.get("Content-Length"):
        return 400, "INVALID_REQUEST", "Content-Length and Transfer-Encoding cannot be combined."
    if transfer_encoding:
        return 400, "UNSUPPORTED_TRANSFER_ENCODING", "Transfer-Encoding is not supported."
    return None


def cross_origin_error(handler):
    base_url = handler.server.store.base_url
    origin = handler.headers.get("Origin")
    if origin and origin != base_url:
        return 403, "ORIGIN_NOT_ALLOWED", "Origin is not allowed."
    if referer := handler.headers.get("Referer"):
        parsed = urlparse(referer)
        if not parsed.scheme or not parsed.netloc or f"{parsed.scheme}://{parsed.netloc}" != base_url:
            return 403, "ORIGIN_NOT_ALLOWED", "Referer is not allowed."
    if handler.headers.get("Sec-Fetch-Site") not in {None, "same-origin", "none"}:
        return 403, "ORIGIN_NOT_ALLOWED", "Sec-Fetch-Site is not allowed."
    return None


def bearer_token(handler):
    value = handler.headers.get("Authorization", "")
    return value.removeprefix("Bearer ") if value.startswith("Bearer ") else ""


def auth_api(handler):
    token = bearer_token(handler)
    if not token:
        handler.fail(401, "UNAUTHORIZED", "Bearer token is required.")
        return False
    if not safe_equal(token, handler.server.api_token):
        handler.fail(403, "FORBIDDEN", "Token is not allowed for this endpoint.")
        return False
    return True


def auth_capture(handler, capture, scope):
    token = bearer_token(handler)
    if not token:
        handler.fail(401, "UNAUTHORIZED", "Bearer token is required.")
        return None
    if scope in {"api", "status", "download"} and safe_equal(token, handler.server.api_token):
        return "api"
    if scope in {"bridge", "status", "download"} and safe_equal(token, capture["bridgeToken"]):
        return "bridge"
    handler.fail(403, "FORBIDDEN", "Token is not allowed for this endpoint.")
    return None


def rate_limited(handler, token, bucket_name, now_window):
    key = f"{token}:{bucket_name}"
    with handler.server.rate_lock:
        bucket = handler.server.rate_buckets.get(key)
        if bucket is None or bucket["window"] != now_window:
            handler.server.rate_buckets[key] = {"window": now_window, "count": 1}
            return False
        if bucket["count"] >= handler.server.rate_limits[bucket_name]:
            return True
        bucket["count"] += 1
        return False
